### PR TITLE
Steps object

### PR DESCRIPTION
**Pull request by gilemos**
Hey everyone!
So, I made a commute and steps object that has the data of a commute from place A to place B. The commute gives basic info about the trip, such as time and spatial distance, and has an array of "Steps", which contains info about how to actually go to one place to the other. Each step has a distance, a time, a vehicle (train, bus, etc) - which is a dictionary with the name of the vehicle, the line, the contact information of the company, etc - direction to go, line, departure time, arrival time, number of stops, among others. The Commute Dictionary looks like this:
Screen Shot 2019-07-20 at 4 06 09 PM

### CODE REVIEW

**Comment by gilemos** [comment]
Observation: some places do not have some info, such as the cost of the trip (like the example above). In those cases the values will be null or -1

File: Travel Scheduler/APIManager.m [hunk 1]
```diff
@@ -191,6 +186,34 @@ -(void)getDistanceFromOrigin:(NSString *)origin toDestination:(NSString *)destin
     [task resume];
 }
 
+//Departure time must be an integer in seconds since midnight, January 1, 1970 UTC
```
Comment (ArunavhaChanda) [inline Travel Scheduler/APIManager.m:183]: This is an example of a good comment.

File: Travel Scheduler/APIManager.m [hunk 2]
```diff
@@ -191,6 +186,34 @@ -(void)getDistanceFromOrigin:(NSString *)origin toDestination:(NSString *)destin
...
+-(void)getCommuteDetailsFromOrigin:(NSString *)originId toDestination:(NSString *)destinationId withDepartureTime:(int)departureTime withCompletion:(void (^)(NSArray *commuteDetailsArray, NSError *error))completion {
+    NSString *urlString = [NSString stringWithFormat:@"https://maps.googleapis.com/maps/api/directions/json?origin=place_id:%@&destination=place_id:%@&mode=transit&departure_time=%d&transit_mode=train|tram|subway|bus&key=%@",originId,destinationId,departureTime,consumerKey];
+    NSURL *url = [NSURL URLWithString:[urlString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]]];
+    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
+    NSURLRequest *request = [NSURLRequest requestWithURL:url];
```
Comment (ArunavhaChanda) [inline Travel Scheduler/APIManager.m]: Lines 191-194: Function?

File: Travel Scheduler/Test/APITesting.m [hunk 3]
```diff
@@ -8,6 +8,8 @@
```
Comment (ArunavhaChanda) [inline Travel Scheduler/Test/APITesting.m:8]: In general, for this file, if this is a unit testing initiative, I absolutely LOVE it.
Possibly modifying it such that it has asserts would be a great move. So if you have something like:

`if (x) {`
`    NSLog(@"Yes");`
`} else {`
`    NSLog(@"No");`
`}`

Replace with:

`<assert_method>(x)`

File: Travel Scheduler/Test/Commute.m [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    self = [self init];
+    self.departureTimeInSecondsSince1970 = departureTime;
+    [[APIManager shared]getCommuteDetailsFromOrigin:originId toDestination:destinationId withDepartureTime:departureTime withCompletion:^(NSArray *commuteInfoArray, NSError *error) {
+        if(commuteInfoArray) {
+            [self initAllProperties];
+            [self buildCommuteWithDictionary:commuteInfoArray[0]];
+            [self makeArrayOfStepsWithRootDictionary:commuteInfoArray[0]];
+        }
+        else {
+            NSLog(@"did not work snif");
```
Comment (ArunavhaChanda) [inline Travel Scheduler/Test/Commute.m:25]: So in final versions of code, you shouldn't have NSLogs. Instead, consider how you can have either fallbacks or failure handling in the most graceful way possible. Depending on the scenario, it can be a default value, re-trying, informing the user and requesting again or even sending back to home screen.
These decisions are what drives user experience and edge case designs for best product quality, and worth having a group discussion about!

File: Travel Scheduler/Test/Commute.m [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        }
+    }];
+    return self;
+}
+
+-(void)initAllProperties {
```
Comment (ArunavhaChanda) [inline Travel Scheduler/Test/Commute.m]: I would avoid names with `init` unless they are public initialization APIs. Name it something else like `createAllProperties` or anything else that doesn't have `init`

File: Travel Scheduler/Test/Step.m [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        self.secondsBetweenTwoDepartures = rootDictionary[@"transit_details"][@"headway"];
+    }
+    else {
+        self.secondsBetweenTwoDepartures = [[NSNumber alloc]initWithInt:-1];
+    }
+    
+    return self;
+}
+
+-(void)initAllProperties {
```
Comment (ArunavhaChanda) [inline Travel Scheduler/Test/Step.m]: Ditto

**Review by ArunavhaChanda — COMMENTED** [review]
I'll leave for Angela/Franklin to accept feature-wise, but those were my comments. None hard blockers, but nice to have!

File: Travel Scheduler/Test/Commute.m (hunk 4 above)
Comment (frankboamps) [reply Travel Scheduler/Test/Commute.m:25]: Yeah. I was really confused on what we should put at those locations when we plan on removing the NSLogs. Thanks for the clarification. We would do well to discuss it before making final version of code.

File: Travel Scheduler/Test/Commute.m [hunk 7]
```diff
@@ -13,11 +13,11 @@
 @implementation Commute
 
 -(instancetype)initWithOrigin:(NSString *)originId toDestination:(NSString *)destinationId withDepartureTime:(int)departureTime{
-    self = [self init];
+    self = [super init];
```
Comment (frankboamps) [inline Travel Scheduler/Test/Commute.m:16]: Good catch here. Nice updates!
